File: backend/internal/grpc/kernel/ipc.go
package kernel

import (
	pb "github.com/GriffinCanCode/AgentOS/backend/proto/kernel"
)

// buildIPCSyscall builds IPC syscall requests (pipes and shared memory)
func (k *KernelClient) buildIPCSyscall(req *pb.SyscallRequest, syscallType string, params map[string]interface{}) {
	switch syscallType {
	// IPC - Pipes
	case "create_pipe":
		readerPid, _ := params["reader_pid"].(uint32)
		writerPid, _ := params["writer_pid"].(uint32)
		createPipeCall := &pb.CreatePipeCall{
			ReaderPid: readerPid,
			WriterPid: writerPid,
		}
		if capacity, hasCapacity := params["capacity"].(uint32); hasCapacity {
			createPipeCall.Capacity = &capacity
		}
		req.Syscall = &pb.SyscallRequest_CreatePipe{
			CreatePipe: createPipeCall,
		}
	case "write_pipe":
		pipeId, _ := params["pipe_id"].(uint32)
		data, _ := params["data"].([]byte)
		req.Syscall = &pb.SyscallRequest_WritePipe{
			WritePipe: &pb.WritePipeCall{PipeId: pipeId, Data: data},
		}
	case "read_pipe":
		pipeId, _ := params["pipe_id"].(uint32)
		size, _ := params["size"].(uint32)
		req.Syscall = &pb.SyscallRequest_ReadPipe{
			ReadPipe: &pb.ReadPipeCall{PipeId: pipeId, Size: size},
		}
	case "close_pipe":
		pipeId, _ := params["pipe_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_ClosePipe{
			ClosePipe: &pb.ClosePipeCall{PipeId: pipeId},
		}
	case "destroy_pipe":
		pipeId, _ := params["pipe_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_DestroyPipe{
			DestroyPipe: &pb.DestroyPipeCall{PipeId: pipeId},
		}
	case "pipe_stats":
		pipeId, _ := params["pipe_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_PipeStats{
			PipeStats: &pb.PipeStatsCall{PipeId: pipeId},
		}
	// IPC - Shared Memory
	case "create_shm":
		size, _ := params["size"].(uint32)
		req.Syscall = &pb.SyscallRequest_CreateShm{
			CreateShm: &pb.CreateShmCall{Size: size},
		}
	case "attach_shm":
		segmentId, _ := params["segment_id"].(uint32)
		readOnly, _ := params["read_only"].(bool)
		req.Syscall = &pb.SyscallRequest_AttachShm{
			AttachShm: &pb.AttachShmCall{SegmentId: segmentId, ReadOnly: readOnly},
		}
	case "detach_shm":
		segmentId, _ := params["segment_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_DetachShm{
			DetachShm: &pb.DetachShmCall{SegmentId: segmentId},
		}
	case "write_shm":
		segmentId, _ := params["segment_id"].(uint32)
		offset, _ := params["offset"].(uint32)
		data, _ := params["data"].([]byte)
		req.Syscall = &pb.SyscallRequest_WriteShm{
			WriteShm: &pb.WriteShmCall{SegmentId: segmentId, Offset: offset, Data: data},
		}
	case "read_shm":
		segmentId, _ := params["segment_id"].(uint32)
		offset, _ := params["offset"].(uint32)
		size, _ := params["size"].(uint32)
		req.Syscall = &pb.SyscallRequest_ReadShm{
			ReadShm: &pb.ReadShmCall{SegmentId: segmentId, Offset: offset, Size: size},
		}
	case "destroy_shm":
		segmentId, _ := params["segment_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_DestroyShm{
			DestroyShm: &pb.DestroyShmCall{SegmentId: segmentId},
		}
	case "shm_stats":
		segmentId, _ := params["segment_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_ShmStats{
			ShmStats: &pb.ShmStatsCall{SegmentId: segmentId},
		}
	// IPC - Memory-Mapped Files
	case "mmap":
		path, _ := params["path"].(string)
		offset, _ := params["offset"].(uint32)
		length, _ := params["length"].(uint32)
		prot, _ := params["prot"].(uint32)
		shared, _ := params["shared"].(bool)
		req.Syscall = &pb.SyscallRequest_Mmap{
			Mmap: &pb.MmapCall{
				Path:   path,
				Offset: offset,
				Length: length,
				Prot:   prot,
				Shared: shared,
			},
		}
	case "mmap_read":
		mmapId, _ := params["mmap_id"].(uint32)
		offset, _ := params["offset"].(uint32)
		length, _ := params["length"].(uint32)
		req.Syscall = &pb.SyscallRequest_MmapRead{
			MmapRead: &pb.MmapReadCall{
				MmapId: mmapId,
				Offset: offset,
				Length: length,
			},
		}
	case "mmap_write":
		mmapId, _ := params["mmap_id"].(uint32)
		offset, _ := params["offset"].(uint32)
		data, _ := params["data"].([]byte)
		req.Syscall = &pb.SyscallRequest_MmapWrite{
			MmapWrite: &pb.MmapWriteCall{
				MmapId: mmapId,
				Offset: offset,
				Data:   data,
			},
		}
	case "msync":
		mmapId, _ := params["mmap_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_Msync{
			Msync: &pb.MsyncCall{MmapId: mmapId},
		}
	case "munmap":
		mmapId, _ := params["mmap_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_Munmap{
			Munmap: &pb.MunmapCall{MmapId: mmapId},
		}
	case "mmap_stats":
		mmapId, _ := params["mmap_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_MmapStats{
			MmapStats: &pb.MmapStatsCall{MmapId: mmapId},
		}
	// IPC - Async Queues
	case "create_queue":
		queueType, _ := params["queue_type"].(string)
		createQueueCall := &pb.CreateQueueCall{
			QueueType: queueType,
		}
		if capacity, hasCapacity := params["capacity"].(uint32); hasCapacity {
			createQueueCall.Capacity = &capacity
		}
		req.Syscall = &pb.SyscallRequest_CreateQueue{
			CreateQueue: createQueueCall,
		}
	case "send_queue":
		queueId, _ := params["queue_id"].(uint32)
		data, _ := params["data"].([]byte)
		sendQueueCall := &pb.SendQueueCall{
			QueueId: queueId,
			Data:    data,
		}
		if priority, hasPriority := params["priority"].(uint32); hasPriority {
			sendQueueCall.Priority = &priority
		}
		req.Syscall = &pb.SyscallRequest_SendQueue{
			SendQueue: sendQueueCall,
		}
	case "receive_queue":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_ReceiveQueue{
			ReceiveQueue: &pb.ReceiveQueueCall{QueueId: queueId},
		}
	case "subscribe_queue":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_SubscribeQueue{
			SubscribeQueue: &pb.SubscribeQueueCall{QueueId: queueId},
		}
	case "unsubscribe_queue":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_UnsubscribeQueue{
			UnsubscribeQueue: &pb.UnsubscribeQueueCall{QueueId: queueId},
		}
	case "close_queue":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_CloseQueue{
			CloseQueue: &pb.CloseQueueCall{QueueId: queueId},
		}
	case "destroy_queue":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_DestroyQueue{
			DestroyQueue: &pb.DestroyQueueCall{QueueId: queueId},
		}
	case "queue_stats":
		queueId, _ := params["queue_id"].(uint32)
		req.Syscall = &pb.SyscallRequest_QueueStats{
			QueueStats: &pb.QueueStatsCall{QueueId: queueId},
		}
	}
}

File: kernel/src/syscalls/types/syscall/ipc.rs
/*!
 * IPC Syscalls
 * Inter-process communication operations (pipes, shared memory, queues, mmap)
 */

use crate::core::types::{Pid, Size};
use serde::{Deserialize, Serialize};

/// IPC operations
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case", tag = "syscall")]
#[non_exhaustive]
pub enum IpcSyscall {
    // ========================================================================
    // Pipes
    // ========================================================================

    /// Create pipe for IPC
    CreatePipe {
        /// Reader process ID
        reader_pid: Pid,
        /// Writer process ID
        writer_pid: Pid,
        /// Optional capacity (bytes)
        capacity: Option<Size>,
    },

    /// Write data to pipe
    WritePipe {
        /// Pipe ID
        pipe_id: Pid,
        /// Data to write
        data: Vec<u8>,
    },

    /// Read data from pipe
    ReadPipe {
        /// Pipe ID
        pipe_id: Pid,
        /// Number of bytes to read
        size: Size,
    },

    /// Close pipe end
    ClosePipe {
        /// Pipe ID
        pipe_id: Pid,
    },

    /// Destroy pipe completely
    DestroyPipe {
        /// Pipe ID
        pipe_id: Pid,
    },

    /// Get pipe statistics
    PipeStats {
        /// Pipe ID
        pipe_id: Pid,
    },

    // ========================================================================
    // Shared Memory
    // ========================================================================

    /// Create shared memory segment
    CreateShm {
        /// Size in bytes
        size: Size,
    },

    /// Attach to shared memory segment
    AttachShm {
        /// Segment ID
        segment_id: Pid,
        /// Read-only access
        #[serde(default)]
        read_only: bool,
    },

    /// Detach from shared memory segment
    DetachShm {
        /// Segment ID
        segment_id: Pid,
    },

    /// Write to shared memory
    WriteShm {
        /// Segment ID
        segment_id: Pid,
        /// Offset in bytes
        #[serde(default)]
        offset: usize,
        /// Data to write
        data: Vec<u8>,
    },

    /// Read from shared memory
    ReadShm {
        /// Segment ID
        segment_id: Pid,
        /// Offset in bytes
        #[serde(default)]
        offset: usize,
        /// Number of bytes to read
        size: Size,
    },

    /// Destroy shared memory segment
    DestroyShm {
        /// Segment ID
        segment_id: Pid,
    },

    /// Get shared memory statistics
    ShmStats {
        /// Segment ID
        segment_id: Pid,
    },

    // ========================================================================
    // Memory-Mapped Files
    // ========================================================================

    /// Memory-map a file
    Mmap {
        /// File path to map
        path: String,
        /// Offset in file
        #[serde(default)]
        offset: usize,
        /// Length to map
        length: Size,
        /// Protection flags (read, write, exec as bit flags)
        prot: u8,
        /// Shared (1) or Private (0) mapping
        #[serde(default)]
        shared: bool,
    },

    /// Read from memory-mapped region
    MmapRead {
        /// Mapping ID
        mmap_id: u32,
        /// Offset in mapping
        #[serde(default)]
        offset: usize,
        /// Length to read
        length: Size,
    },

    /// Write to memory-mapped region
    MmapWrite {
        /// Mapping ID
        mmap_id: u32,
        /// Offset in mapping
        #[serde(default)]
        offset: usize,
        /// Data to write
        data: Vec<u8>,
    },

    /// Synchronize mmap to file
    Msync {
        /// Mapping ID
        mmap_id: u32,
    },

    /// Unmap a memory-mapped region
    Munmap {
        /// Mapping ID
        mmap_id: u32,
    },

    /// Get mmap statistics
    MmapStats {
        /// Mapping ID
        mmap_id: u32,
    },

    // ========================================================================
    // Message Queues
    // ========================================================================

    /// Create message queue
    CreateQueue {
        /// Queue type: "fifo", "priority", or "pubsub"
        queue_type: String,
        /// Optional capacity (messages)
        capacity: Option<Size>,
    },

    /// Send message to queue
    SendQueue {
        /// Queue ID
        queue_id: Pid,
        /// Message data
        data: Vec<u8>,
        /// Optional priority (for priority queues)
        priority: Option<u8>,
    },

    /// Receive message from queue
    ReceiveQueue {
        /// Queue ID
        queue_id: Pid,
    },

    /// Subscribe to pubsub queue
    SubscribeQueue {
        /// Queue ID
        queue_id: Pid,
    },

    /// Unsubscribe from pubsub queue
    UnsubscribeQueue {
        /// Queue ID
        queue_id: Pid,
    },

    /// Close queue connection
    CloseQueue {
        /// Queue ID
        queue_id: Pid,
    },

    /// Destroy queue completely
    DestroyQueue {
        /// Queue ID
        queue_id: Pid,
    },

    /// Get queue statistics
    QueueStats {
        /// Queue ID
        queue_id: Pid,
    },
}
